### PR TITLE
Syntax highlighting with CSS classes

### PR DESCRIPTION
**Pull request by uwearzt**
Todo

- [x]  Wait for syntect release (incl. fix for rust-onig)
- [x] Documentation -> @uwearzt 
- [x] Example -> @uwearzt (My Website)
- [ ] Check interface (especially config.toml) @Keats

### CODE REVIEW

**Comment by Keats** [comment]
> Wait for syntect release (incl. fix for rust-onig)

It's been done and I'm currently releasing 0.11 with it. You probably want to cherry-pick your commit on top of current master though.

**Comment by Keats** [comment]
I didn't mean to close all the PRs by deleting the branch :(

**Comment by uwearzt** [comment]
Will rebase over the weekend :), thanks for your patience

**Comment by uwearzt** [comment]
Rebased to next branch, i will create the documentation in the next days.

**Comment by apiraino** [comment]
Hi @Keats I'd like to port at least one Sublime syntax highlight theme to Zola. Do you suggest waiting for this PR to be merged?

thanks

**Comment by Keats** [comment]
> Hi @Keats I'd like to port at least one Sublime syntax highlight theme to Zola. Do you suggest waiting for this PR to be merged?

You can do a PR now

File: components/config/src/config.rs
Comment (Keats) [inline components/config/src/config.rs]: themes

File: components/rendering/src/markdown.rs
Comment (Keats) [inline components/rendering/src/markdown.rs]: can be removed?

File: components/site/src/lib.rs
Comment (Keats) [inline components/site/src/lib.rs]: `generate_highlighting_themes`

File: components/site/src/lib.rs
Comment (Keats) [inline components/site/src/lib.rs]: No need to print.

File: docs/content/documentation/getting-started/configuration.md
Comment (Keats) [inline docs/content/documentation/getting-started/configuration.md]: Hmmm.
I think the inner part is ok, except maybe file -> filename for clarity.

As for `generate_theme_css`, I think `highlighting_themes_css` is more consistent.

**Review by Keats — CHANGES_REQUESTED** [review]
That's a very neat feature!

File: components/config/src/config.rs
Comment (uwearzt) [reply components/config/src/config.rs]: Also renamed the Vector to `generate_themes_css`.

File: components/rendering/src/markdown.rs
Comment (uwearzt) [reply components/rendering/src/markdown.rs]: yes, that was miss on my side

File: components/site/src/lib.rs
Comment (uwearzt) [reply components/site/src/lib.rs]: much better name, thanks!

File: components/site/src/lib.rs
Comment (uwearzt) [reply components/site/src/lib.rs]: Good catch, was in for debugging purposes!

File: docs/content/documentation/getting-started/configuration.md
Comment (uwearzt) [reply docs/content/documentation/getting-started/configuration.md]: Yepp, much better

**Comment by uwearzt** [comment]
Adjusted to Review comments, many thanks for the Review. Should i also squash my commits into one?

**Comment by Keats** [comment]
I've just tried it on my site (https://github.com/Keats/vincentprouillet/) and sadly it panics at that line: https://github.com/trishume/syntect/blob/master/src/parsing/scope.rs#L417
with the following in my config:

```
highlight_code = true
highlight_theme = "css"
highlighting_themes_css = [
  { theme = "cheerfully-light", filename = "cheerfully-light.css" },
]
```
This looks like a bug in syntect, I'll see if I can reproduce it/make a PR with a failing test.

File: components/rendering/src/markdown.rs
Comment (Keats) [inline components/rendering/src/markdown.rs:203]: Is this only highlighting rust..?

File: EXAMPLES.md
Comment (Keats) [inline EXAMPLES.md]: can you revert the changes to that file? I would rather have a link yo your blog source in the docs when talking about dark mode example.

File: components/rendering/src/markdown.rs
Comment (Keats) [reply components/rendering/src/markdown.rs:203]: Maybe the panic is because of that? It's trying to parse a different language than it is

File: EXAMPLES.md
Comment (uwearzt) [reply EXAMPLES.md]: Yes, no problem

File: components/rendering/src/markdown.rs
Comment (uwearzt) [reply components/rendering/src/markdown.rs:203]: At this point i only need a unused syntax highlighter for swapping it out before calling finalize (We discussed that before). 

The highlighter which is used to do the actual highlighting is created in line 239 with the correct syntax set from the codeblock:
```
 match &SYNTAX_SET.find_syntax_by_extension(info) {
```

**Comment by uwearzt** [comment]
Sorry for the delay, i was quite busy. I also build my site with the Syntax highlighting theme you tried without problems. BTW: fixed also a typo inside a new commit. We should squash everything before the merge.

**Comment by uwearzt** [comment]
Tested your site, and got the same panic as you:
```
Building site...
thread 'thread '<unnamed><unnamed>' panicked at '' panicked at 'tried to restore cleared scopes, but none were clearedtried to restore cleared scopes, but none were cleared', ', /Users/uwe/.cargo/registry/src/github.com-1ecc6299db9ec823/syntect-4.2.0/src/parsing/scope.rs/Users/uwe/.cargo/registry/src/github.com-1ecc6299db9ec823/syntect-4.2.0/src/parsing/scope.rs::417417:29:
29note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
will try to debug that .

**Comment by uwearzt** [comment]
I investigated a little bit. the first codeblock failing is the json block in 2014-09-19_finding-trending-things.md.

It fails, even if i strip it down to 

```
```json
{
    "size": 0,
    "aggs": {
    }
}
```

If i strip it down even more, it works fine. I created a test inside syntect with this json which works fine. So a possible culprit is the loaded SyntaxSet. If the loaded SyntaxSet is changed to the SyntaxSet::load_defaults_newlines(); everything work fine without problems.

**Comment by Keats** [comment]
Did you manage to reproduce a failing test for syntect?

**Comment by uwearzt** [comment]
Sorry, i was on vacation, i will try it in the next days.

**Comment by mre** [comment]
@uwearzt any updates on this? Would love to use that for my blog as well. 😊

**Comment by Keats** [comment]
Oops, didn't mean to close that sorry

**Comment by evan-brass** [comment]
I'm still learning about Syntect and Zola, but I don't think that this panic is an issue in either of them. I think it's a problem in the JSON syntax definition within Zola's syntax set.  ClassedHTMLGenerator doesn't panic on that JSON snippet for either of the newlines or nonewlines default syntax sets within Syntect, but it does when using the Zola syntax set (which I believe is newlines).

The Syntect default JSON syntax doesn't produce any clear or restore stack operations while the one in Zola produces a restore operation on a line that it didn't produce a clear for.  There is an unmatched clear from a previous line, though, so if that's supposed to work then I guess there's a problem with the `tokens_to_classed_spans` function which creates a new ScopeStack for each line (losing any previous clears).

https://github.com/evan-brass/zola/blob/752b6c38ff195a78a377b067c5065d6a2eac65cf/components/config/src/highlighting.rs#L66

No matter what though, until trishume/syntect#307 is merged, I don't think ClassedHTMLGenerator will work for Zola because of the newlines Syntax set.

**Comment by uwearzt** [comment]
Sorry, was busy doing some other stuff. I will rebase the PR to the newest next branch now. Then i will try to fix the issues.

**Comment by Keats** [comment]
There is https://github.com/getzola/zola/pull/1242 now that supersedes this PR I think

**Comment by Ahmedss1857** [comment]
{
‏  "ns": "yt",
‏  "el": "detailpage",
‏  "cpn": "WNSske0ngxX-1GO5",
‏  "docid": "RawnOB40IVk",
‏  "ver": 2,
‏  "cmt": "98.991",
‏  "ei": "Wc5EYLWrII6YWLGKt9AO",
‏  "fmt": "18",
‏  "fs": "0",
‏  "rt": "472.896",
‏  "of": "YUH5ANIs5L0RUKzkCJgcpw",
‏  "euri": "",
‏  "lact": 2,
‏  "cl": "360985524",
‏  "mos": 0,
‏  "state": "8",
‏  "vm": "CAMQARgBKixpZ2UyNGlsMjVsZm9tLU5MY1BGcXRCRlJ5eHdGN09FbG45QUVmbWlnbGpJPToyQU9HdF9PTHh1NmtqTHBqMms1UGIzTE56T09yUVVvd1JySnJiTVhuazJzUUE5UFZwbnc",
‏  "volume": 100,
‏  "cbrand": "apple",
‏  "cbr": "Safari Mobile",
‏  "cbrver": "14.0.3.15E148",
‏  "c": "MWEB",
‏  "cver": "2.20210224.03.00",
‏  "cplayer": "UNIPLAYER",
‏  "cmodel": "iphone",
‏  "cos": "iPhone",
‏  "cosver": "14_4",
‏  "cplatform": "MOBILE",
‏  "hl": "ar_EG",
‏  "cr": "SA",
‏  "len": "493.053",
‏  "fexp":[]()

**Comment by Ahmedss1857** [comment]
{
‏  "ns": "yt",
‏  "el": "detailpage",
‏  "cpn": "WNSske0ngxX-1GO5",
‏  "docid": "RawnOB40IVk",
‏  "ver": 2,
‏  "cmt": "98.991",
‏  "ei": "Wc5EYLWrII6YWLGKt9AO",
‏  "fmt": "18",
‏  "fs": "0",
‏  "rt": "472.896",
‏  "of": "YUH5ANIs5L0RUKzkCJgcpw",
‏  "euri": "",
‏  "lact": 2,
‏  "cl": "360985524",
‏  "mos": 0,
‏  "state": "8",
‏  "vm": "CAMQARgBKixpZ2UyNGlsMjVsZm9tLU5MY1BGcXRCRlJ5eHdGN09FbG45QUVmbWlnbGpJPToyQU9HdF9PTHh1NmtqTHBqMms1UGIzTE56T09yUVVvd1JySnJiTVhuazJzUUE5UFZwbnc",
‏  "volume": 100,
‏  "cbrand": "apple",
‏  "cbr": "Safari Mobile",
‏  "cbrver": "14.0.3.15E148",
‏  "c": "MWEB",
‏  "cver": "2.20210224.03.00",
‏  "cplayer": "UNIPLAYER",
‏  "cmodel": "iphone",
‏  "cos": "iPhone",
‏  "cosver": "14_4",
‏  "cplatform": "MOBILE",
‏  "hl": "ar_EG",
‏  "cr": "SA",
‏  "len": "493.053",
‏  "fexp":

**Pull request closed without merging**
